### PR TITLE
Add a deprecation DSL

### PR DESCRIPTION
**Pull request by bf4**
Follows https://github.com/rails-api/active_model_serializers/pull/1543
and 6b4c8df6fb6bc142ee6a74da51bb26c42a025b3c

### CODE REVIEW

**Comment by bf4** [comment]
cc @groyoh

File: lib/active_model/serializer/adapter.rb [hunk 1]
```diff
@@ -3,40 +3,37 @@ class Serializer
     # @deprecated Use ActiveModelSerializers::Adapter instead
     module Adapter
       class << self
+        extend ActiveModelSerializers::Deprecate
+
         def create(resource, options = {})
```
Comment (groyoh) [inline lib/active_model/serializer/adapter.rb]: Would `delegate :create, to: ActiveModelSerializers::Adapter` work here?

**Comment by groyoh** [comment]
Awesome work :100:

File: lib/active_model/serializer/adapter.rb (hunk 1 above)
Comment (bf4) [inline lib/active_model/serializer/adapter.rb]: y'know, I tried it with ruby's delegate and forwardable, but not with rails' delegate.  Feel free to try. I think you can push to my branch.

**Comment by bf4** [comment]
@groyoh to be honest, I'm not sure if I like this as a 'another thing in AMS. OTOH, it does take most of the drudgery out of deprecating. OTOH (three hands), we can remove it when we hit 0.10.0.

**Comment by groyoh** [comment]
I also think that it would be better to get rid of it once we hit 0.10.0. But right now, it definitely makes things cleaner and easier to handle, especially once we move `ActiveModel::Serializer` to `ActiveModelSerializers` module.

File: lib/active_model/serializer/adapter.rb [hunk 2]
```diff
@@ -3,40 +3,37 @@ class Serializer
...
         def lookup(adapter)
-          warn_deprecation
           ActiveModelSerializers::Adapter.lookup(adapter)
         end
-
-        def warn_deprecation
-          warn "Calling deprecated #{name} (#{__FILE__}) from #{caller[1..3].join(', ')}. Please use ActiveModelSerializers::Adapter"
-        end
-        private :warn_deprecation
+        deprecate :lookup, 'ActiveModelSerializers::Adapter.'
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/adapter.rb]: I really like this implementation, good thinkin'

File: lib/active_model_serializers/deprecate.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#       end
+#     end
+#
+# Adapted from https://github.com/rubygems/rubygems/blob/1591331/lib/rubygems/deprecate.rb
+module ActiveModelSerializers
+  module Deprecate
+    ##
+    # Simple deprecation method that deprecates +name+ by wrapping it up
+    # in a dummy method. It warns on each call to the dummy method
+    # telling the user of +replacement+ (unless +replacement+ is :none) that it is planned to go away.
```
Comment (NullVoxPopuli) [inline lib/active_model_serializers/deprecate.rb]: good docs!

File: lib/active_model_serializers/deprecate.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+#       end
+#
+#       extend ActiveModelSerializers::Deprecate
+#       deprecate :instance_method, "ActiveModelSerializers::NewPlace#new_method"
+#
+#       class << self
+#         extend ActiveModelSerializers::Deprecate
+#         deprecate :klass_method, :none
+#       end
+#     end
```
Comment (NullVoxPopuli) [inline lib/active_model_serializers/deprecate.rb]: I like the example!

**Comment by NullVoxPopuli** [comment]
I actually think we should keep the deprecation module when we hit 0.10. That way when can use the same code if nothing better comes around by the time we need to start deprecating stuff again.

Unless there are any more TODOs, I'm good with merging this.

**Comment by groyoh** [comment]
@NullVoxPopuli Wouldn't deprecation lead to a breaking change at that point?

**Comment by NullVoxPopuli** [comment]
yeah, I mean, remove the deprecations when we hit 0.10, but keep the ability to deprecate with `ActiveModelSerializers::Deprecate`

**Comment by NullVoxPopuli** [comment]
Actually, we still need a changelog in here -> then I'm ok with merging.

**Comment by NullVoxPopuli** [comment]
closed in favor of #1559

File: lib/active_model_serializers/deprecate.rb (hunk 3 above)
Comment (bf4) [inline lib/active_model_serializers/deprecate.rb]: Copied

**Pull request closed without merging**